src/app/reporting/SynchronizedReportSchedulerImpl.h

```diff
@@ -30,6 +30,48 @@ using Milliseconds64  = System::Clock::Milliseconds64;
 using ReadHandlerNode = ReportScheduler::ReadHandlerNode;
 using TimerDelegate   = ReportScheduler::TimerDelegate;
 
+/**
+ * @class Synchronized ReportSchedulerImpl
+ *
+ * @brief This class extends ReportSchedulerImpl and overrides it's scheduling logic.
+ *
+ * It only overrides Observers method where the scheduling logic make it necessary, the others are kept as is.
+ *
+ * It inherits from TimerContext so that it can be used as a TimerDelegate instead on relying on the nodes to schedule themselves.
+ *
+ * ## Scheduling Logic
+ *
+ * This class implements a scheduling logic that aims to make all ReadHandlers report at the same time when possible.
+ * The goal is to minimize the different times a device wakes up to report, and thus this aims to schedule all reports at the latest
+ * possible time while ensuring that all reports get sent before their max interval.
+ *
+ * The logic also aims to minimize the impact on the responsivity of the device.
+ *
+ * The scheduling logic is as follows:
+ * - The CalculateNextReportTimeout is called by the same ReadHandler Observer callbacks than the non-synchronized implementation:
```

Comment:
I think the key part here is that whenever any state that might affect when/whether reporting should happen changes we call CalculateNextReportTimeout, right?  Making that clear (and maybe giving those callbacks as examples of such state changes) would be much better.